[ADD] Studio: New webhooks doc (#12255)
odoo/documentation: content/applications/studio/studio_webhooks.rst

Set the webhook's trigger
~~~~~~~~~~~~~~~~~~~~~~~~~

To set up this webhook, open the **Sales** app. Then, follow the steps above for setting up a
webhook's trigger while making sure the :guilabel:`Model` is set to `Sales Order`. Also, set the
:guilabel:`Target Record` to `model.env[payload.get('model')].browse(int(payload.get('id')))`. This
is broken down below.

- **model**: What gets updated in Odoo (in this case, a sales order). This will match the
  :guilabel:`Model` set earlier.
- **env**: Where the action takes place. In this case, it is Odoo.
- **payload**: This is what is sent to the webhook's URL. It will contain the information that
  updates the sales order.
- **get('model')**: Tells the webhook what database record to look at. In this case, the webhook
  will retrieve (`get`) the data tied to a specific record `model`. In this example, this is the
  sales orders.
- **browse**: This tells the webhook to look in the `model` (the sales orders) set by the payload
  for what to update.
- **int**: This will turn the target into an `integer` (a whole number). This is important in case
  some words (a `string`) or a decimal number is included in the payload's target record.

> - **get('id')**: Tells the webhook what to do with the payload. In this case, the webhook will
>   retrieve (`get`) the data tied to a specific record `id`. In this example, this is the sales order
>   number that will get updated in Odoo.

Reviewer: ```suggestion
- **get('id')**: Identifies the sales order number that is getting updated in Odoo.
```

Set the webhook's action
~~~~~~~~~~~~~~~~~~~~~~~~

> Once the trigger is set, the webhook's action needs to be set. Begin by clicking :guilabel:`Add an
> action`. For the :guilabel:`Type`, click :guilabel:`Update Record`. And then, set the

Reviewer: ```suggestion
After setting the trigger, set the webhook's action by clicking :guilabel:`Add an
action`. For the :guilabel:`Type`, click :guilabel:`Update Record`. And then, set the
```

:guilabel:`Action Details` to `Update Customer to USD`. Finally, click :guilabel:`Save & Close`.

Webhook setup summary
~~~~~~~~~~~~~~~~~~~~~

> To summarize what is set up, the webhook will target sales orders, identified by their sales order

Reviewer: For brevity, consider removing "will" where it makes sense throughout the doc (:

before the number. For example, a sales order with the number `S00007` should be entered as `7` when
it is entered into Postman. Finally, click :guilabel:`Send` in Postman.

If a message saying `200 OK` or `status: ok` gets returned, then the webhook is functioning properly
on Odoo's side. From here, implementation can begin with the other tool to automatically send those
webhook calls into Odoo using the webhook's URL.

If any other responses are returned, the number associated with them will help to identify the
problem. For example, a `500 Internal Server Error` means that Odoo could not interpret the call
properly. If this gets returned, ensure the `model` and `id` fields are properly mapped in the
webhook's configuration and in Postman.

Create a new contact
--------------------

This webhook will create a new contact in an Odoo database. This could be helpful for automatically
creating new vendors or customers. This webhook will use custom code, so consult with a developer to
ensure the code matches any unique setups a database might have.

Set the webhooks' trigger
~~~~~~~~~~~~~~~~~~~~~~~~~

> To set up this webhook, open the **Contacts** app. Then, follow the steps above for setting up a

Reviewer: In general, we avoid saying "follow the steps above" since we assume readers skip around the doc. Instead, let's add a custom anchor to the webhook setup instructions, and link it here, instead saying something like "Follow the steps in the :ref:`Create a studio webhook <custom_anchor>` section to do blabala"

action`. For the :guilabel:`Type`, click :guilabel:`Execute Code`. And then, set the
:guilabel:`code` to the sample code below. Finally, click :guilabel:`Save & Close`.

.. code-block:: python

   try:  # to retrieve data from the payload
       data = request.get_json_data()
   except:
       data = {}

   # variables to hold data from the payload contact_name = data.get('name') contact_email =
   data.get('email') contact_phone = data.get('phone')

   # a Python function to turn the payload into a contact in Odoo if contact_name and contact_email:
       new_partner = env['res.partner'].create({
           'name': contact_name, 'email': contact_email, 'phone': contact_phone, 'company_type':
           'person', 'customer_rank': 1,
       })

       env['ir.logging'].create({  # for error logging
           'name': 'Webhook Contact Creation', 'type': 'server', 'level': 'info', 'message':
           f"Created new contact: {new_partner.name} (ID: {new_partner.id})", 'path': 'webhook',
           'func': 'execute', 'line': 0,
       })
   else:  # an error message for missing required data in the payload

>        raise ValueError("Missing required fields: 'name' and 'email'")

Reviewer: Amazing job obtaining this code snippet! The formatting looks great too!

Turn on developer mode
----------------------

Developer mode must be turned on to set what kinds of database records the webhook will target (like
sales orders or contact information). Learn how to :doc:`turn on and off developer
mode <../general/developer_mode>`.

Set the webhook's trigger
-------------------------

To create a **Studio** webhook within another app, click the :icon:`oi-studio` (**Studio**) icon
will appear next to the profile photo. Then, at the top of the screen, click :guilabel:`Webhooks`,
and then select :guilabel:`New`. From here, name the webhook, set the webhook's model (what kind of
database entry gets targeted), and toggle whether calls made to the
webhook URL should be logged (which would track the webhook's call history). Finally, adjust the
:guilabel:`Target Record` actions if the webhook will accept payloads (i.e., message content) from a
non-Odoo source.

.. danger::
   The webhook's URL is **confidential** and should be treated with care. Sharing it online or
   without caution can provide unintended access to the Odoo database. Click :guilabel:`Rotate
   Secret` to change the URL if needed.

Set the webhook's action
------------------------

> Next, set the webhook's actions. This is what will happen when the webhook's URL is called. Remember

Reviewer: > Next, set the webhook's actions.

Where in the UI do we do this? In the Actions to Do tab, click what, and that opens what window, from there, then what? Let's also document what all this means 

![image](https://github.com/user-attachments/assets/04afac38-aa3b-49a2-a1a9-4c54b3758ad0)
Since I know you go into the details later in the examples, we just need a walkthrough and maybe link to specific examples using a [ref](https://www.odoo.com/documentation/18.0/contributing/documentation/rst_guidelines.html#custom-anchors)

that the payload (the contents of the webhook) may dictate what the webhook's actions do, so pay

> attention to how the actions are set up so the external tool can also be properly set up to provide

Reviewer: > so pay
attention to how the actions are set up so the external tool can also be properly set up to provide...

This section could be more specific—what fields should we focus on? Right now, webhooks feel a bit abstract without an example. We could either elaborate on all the relevant fields or leave this out entirely since, as a reader, it's unclear what to pay attention to.

If this ties into webhooks, could we clarify what Update vs. Compute do? Instead of "pay attention to how actions are set up," we might say something more actionable, like:

> Ensure the correct Odoo fields are targeted, whether you're updating or computing, and that the action type is configured correctly.

Also, what happens if the action isn't set up correctly? Is there a way to check at this stage?

the right information in the payload.

.. important::
   Although the model is set in Odoo, its payload format differs and must be included in the

>    payload. Click the model's internal link to find the correct format in the Model field. For

Reviewer: guilabel for model!

   example, a sales order webhook uses the `Sales Order` model but has a `sale.order` payload
   format.

   .. image:: studio_webhooks/webhook-sales-order-model.png
      :alt: The sales order model for a webhook's payload.

Webhook use cases
=================

Below are two examples of how to use webhooks in Odoo. These webhooks require external tools (which
are listed with the example).

.. warning::
   Consult with a developer, solution architect, or another technical role when deciding to
   implement webhooks. If not properly configured, webhooks may disrupt the Odoo database and will
   take time to revert.

Update a sales order's currency
-------------------------------

> This webhook updates a sales order in the **Sales** app to USD, useful for subsidiaries outside the
> United States with a mother company located inside the United States or during mergers when
> consolidating data into one Odoo database.

Reviewer: optional for brevity
```suggestion
This webhook updates a sales order's currency to USD  in the **Sales** app, useful for subsidiaries of a
U.S.-based parent company or during mergers consolidating data into one Odoo database.
```

Set the webhook's trigger
~~~~~~~~~~~~~~~~~~~~~~~~~

To set up this webhook, open the **Sales** app. Then, follow the steps above for setting up a

> webhook's trigger while making sure the :guilabel:`Model` is set to `Sales Order`. Also, set the

Reviewer: guilabels for elements that appear on the UI. Typed user inputs like product name, name of the model, are the only times we format with simply back ticks (`)
```suggestion
webhook's trigger while making sure the :guilabel:`Model` is set to :guilabel:`Sales Order`. Also, set the
```

:guilabel:`Target Record` to `model.env[payload.get('model')].browse(int(payload.get('id')))`. This
is broken down below.

- **model**: What gets updated in Odoo (in this case, a sales order). This will match the
  :guilabel:`Model` set earlier.
- **env**: Where the action takes place. In this case, it is Odoo.
- **payload**: This is what is sent to the webhook's URL. It will contain the information that
  updates the sales order.
- **get('model')**: Tells the webhook what database record to look at. In this case, the webhook
  will retrieve (`get`) the data tied to a specific record `model`. In this example, this is the
  sales orders.
- **browse**: This tells the webhook to look in the `model` (the sales orders) set by the payload
  for what to update.

> - **int**: This will turn the target into an `integer` (a whole number). This is important in case
>   some words (a `string`) or a decimal number is included in the payload's target record.

Reviewer: Include only if it's true
```suggestion
- **int**: Converts the target into an `integer` (a whole number). This is important in case
  the data received is in a different format, like a *string* or a decimal number.
```
